result.go

```diff
@@ -72,19 +70,89 @@ func newResult(t reflect.Type) (result, error) {
 	}
 }
 
+// resultVisitor visits every result in a result tree, allowing tracking state
+// at each level.
+type resultVisitor interface {
+	// Visit is called on the result being visited.
+	//
+	// If Visit returns a non-nil resultVisitor, that resultVisitor visits all
+	// the child results of this result.
+	Visit(result) resultVisitor
+
+	// VisitField is called on each field of a resultObject after visiting it
+	// but before walking its descendants.
+	//
+	// The same resultVisitor is used for all fields: the one returned upon
+	// visiting the resultObject.
+	//
+	// For each visited field, if VisitField returns a non-nil resultVisitor,
+	// it will be used to walk the result of that field.
+	VisitField(resultObjectField) resultVisitor
+
+	// VisitPosition is called with the index of each result of a resultList
+	// after vising it but before walking its descendants.
+	//
+	// The same resultVisitor is used for all results: the one returned upon
+	// visiting the resultList.
+	//
+	// For each position, if VisitPosition returns a non-nil resultVisitor, it
+	// will be used to walk the result at that index.
+	VisitPosition(idx int) resultVisitor
+}
+
+// walkResult walks the result tree for the given result with the provided
+// visitor.
+//
+// resultVisitor.Visit will be called on the provided result and if a non-nil
+// resultVisitor is received, it will be used to walk its descendants. If a
+// resultObject or resultList was visited, VisitField and VisitPosition
+// respectively will be called before visiting the descendants of that
+// resultObject/resultList.
+//
+// This is very similar to how go/ast.Walk works.
+func walkResult(r result, v resultVisitor) {
+	v = v.Visit(r)
+	if v == nil {
+		return
+	}
+
+	switch res := r.(type) {
+	case resultSingle, resultError:
+		// No sub-results
+	case resultObject:
+		w := v
+		for _, f := range res.Fields {
+			if v := w.VisitField(f); v != nil {
+				walkResult(f.Result, v)
+			}
+		}
+	case resultList:
+		w := v
+		for i, r := range res.Results {
+			if v := w.VisitPosition(i); v != nil {
+				walkResult(r, v)
+			}
+		}
+	default:
+		panic(fmt.Sprintf(
+			"It looks like you have found a bug in dig. "+
+				"Please file an issue at https://github.com/uber-go/dig/issues/ "+
```

Comment:
not sure if indentation is intentional...

Reply:
gofmt insists on indenting that